File: packages/android/core-domain/src/main/kotlin/app/notes/domain/CalendarProvider.kt
package app.notes.domain

/** Opaque id for an event on the external provider (Google event.id etc.). */
@JvmInline
value class ExternalEventId(val raw: String) {
    init { require(raw.isNotBlank()) { "ExternalEventId cannot be empty" } }
}

data class DateRange(val start: java.time.Instant, val end: java.time.Instant) {
    init { require(!end.isBefore(start)) { "DateRange.end must be >= start" } }
}

data class NewEventInput(
    val title: String,
    val startAt: java.time.Instant,
    val endAt: java.time.Instant,
    val description: String? = null,
    val attendees: List<String> = emptyList(),
    val reminderMinutes: Int? = null,
    val noteId: NoteId? = null,
) {
    init {
        require(title.isNotBlank()) { "NewEventInput.title cannot be empty" }
        require(!endAt.isBefore(startAt)) { "NewEventInput.endAt must be >= startAt" }
    }
}

data class EventPatch(
    val title: String? = null,
    val startAt: java.time.Instant? = null,
    val endAt: java.time.Instant? = null,
    val description: String? = null,
)

sealed class CalendarChange {
    data class Upsert(val event: CalendarEvent) : CalendarChange()
    data class Delete(val externalEventId: ExternalEventId) : CalendarChange()
}

/**
 * The calendar port. V1 implementation is `GoogleCalendarAdapter`
 * (server-side, TS). V2 swaps in OutlookCalendarAdapter behind the same
 * shape. UI code never imports the adapter; the Edge Functions own all
 * server-side calls.
 */
interface CalendarProvider {
    val id: CalendarProviderId
    suspend fun listEvents(range: DateRange): List<CalendarEvent>
    suspend fun createEvent(input: NewEventInput): CalendarEvent
    suspend fun updateEvent(id: ExternalEventId, patch: EventPatch): CalendarEvent
    suspend fun deleteEvent(id: ExternalEventId)
    /**
     * V1 polling: return the set of CalendarChanges for the range. The
     * Android binding wraps this in a WorkManager periodic worker
     * (15-min cadence while foregrounded). V2 swaps in push channels —
     * the bindings change, the port stays.
     */
    suspend fun pollEvents(range: DateRange): List<CalendarChange>
}

File: packages/domain/src/calendar-provider.ts
import type { NoteId } from "./ids.js";
import type { CalendarEvent } from "./calendar.js";
import type { CalendarProviderId } from "./enums.js";

/** Opaque id for an event on the external provider (e.g. Google's event.id). */
export type ExternalEventId = string & { readonly __brand: "ExternalEventId" };

export function ExternalEventId(value: string): ExternalEventId {
  if (value.trim().length === 0) {
    throw new Error("ExternalEventId cannot be empty");
  }
  return value as ExternalEventId;
}

/** Inclusive date range, ISO-8601 timestamps. */
export interface DateRange {
  readonly start: string;
  readonly end: string;
}

export interface NewEventInput {
  readonly title: string;
  readonly startAt: string;
  readonly endAt: string;
  readonly description?: string;
  readonly attendees?: readonly string[];
  /** Pop-up minutes before the start; the adapter maps to provider semantics. */
  readonly reminderMinutes?: number;
  /** Optional source note for back-reference (M08 note_event_links). */
  readonly noteId?: NoteId;
}

export interface EventPatch {
  readonly title?: string;
  readonly startAt?: string;
  readonly endAt?: string;
  readonly description?: string;
}

/** A single change emitted by `CalendarProvider.watchEvents`. */
export type CalendarChange =
  | { readonly kind: "Upsert"; readonly event: CalendarEvent }
  | { readonly kind: "Delete"; readonly externalEventId: ExternalEventId };

/**
 * The calendar port. V1 implementation is `GoogleCalendarAdapter`; V2 swaps
 * in Outlook/Microsoft 365 behind the same shape. UI code (Android + Web)
 * never imports the adapter; the Edge Functions own all server-side
 * calls.
 */
export interface CalendarProvider {
  readonly id: CalendarProviderId;
  listEvents(range: DateRange): Promise<readonly CalendarEvent[]>;
  createEvent(input: NewEventInput): Promise<CalendarEvent>;
  updateEvent(id: ExternalEventId, patch: EventPatch): Promise<CalendarEvent>;
  deleteEvent(id: ExternalEventId): Promise<void>;
  /**
   * Polled in V1 (15-min cadence while foregrounded); the Push channel
   * binding is V2. The shape is a Flow-equivalent so the V2 swap is a
   * one-file change.
   */
  watchEvents(range: DateRange): AsyncIterable<CalendarChange>;
}
